src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -387,40 +387,54 @@ private static void CreateParentsAndDirectory(string fullPath)
+
             ReadOnlySpan<char> srcNoDirectorySeparator = Path.TrimEndingDirectorySeparator(sourceFullPath.AsSpan());
             ReadOnlySpan<char> destNoDirectorySeparator = Path.TrimEndingDirectorySeparator(destFullPath.AsSpan());
 
+            // When the path ends with a directory separator, it must not be a file.
+            // On Unix 'rename' fails with ENOTDIR, on wasm we need to manually check.
             if (OperatingSystem.IsBrowser() && Path.EndsInDirectorySeparator(sourceFullPath) && FileExists(sourceFullPath))
             {
-                // On Windows we end up with ERROR_INVALID_NAME, which is
-                // "The filename, directory name, or volume label syntax is incorrect."
-                // On Unix, rename fails with ENOTDIR, but on WASM it does not.
-                // So if the path ends with directory separator, but it's a file, we just throw.
                 throw new IOException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
             }
 
-            if (!sameDirectoryDifferentCase) // This check is to allow renaming of directories
+            // The destination must not exist (unless it is a case-sensitive rename).
+            // On Unix 'rename' will overwrite the destination file if it already exists, we need to manually check.
+            if (!isCaseSensitiveRename && Interop.Sys.LStat(destNoDirectorySeparator, out Interop.Sys.FileStatus destFileStatus) >= 0)
             {
-                if (Interop.Sys.Stat(destNoDirectorySeparator, out _) >= 0)
-                {
-                    // destination exists, but before we throw we need to check whether source exists or not
+                // Maintain order of exceptions as on Windows.
 
-                    // Windows will throw if the source file/directory doesn't exist, we preemptively check
-                    // to make sure our cross platform behavior matches .NET Framework behavior.
-                    if (Interop.Sys.Stat(srcNoDirectorySeparator, out Interop.Sys.FileStatus sourceFileStatus) < 0)
-                    {
-                        throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
-                    }
-                    else if ((sourceFileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) != Interop.Sys.FileTypes.S_IFDIR
-                        && Path.EndsInDirectorySeparator(sourceFullPath))
+                // Throw if the source doesn't exist.
+                if (Interop.Sys.LStat(srcNoDirectorySeparator, out Interop.Sys.FileStatus sourceFileStatus) < 0)
+                {
+                    throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
+                }
+                // Source and destination must not be the same file unless it is a case-sensitive rename.
+                else if (sourceFileStatus.Dev == destFileStatus.Dev &&
+                         sourceFileStatus.Ino == destFileStatus.Ino)
+                {
```

Comment:
This is where we now allow renaming from `/a/foo` to /a/Foo` on case-sensitive systems like Linux and FreeBSD when we see both paths are the same file (dev and ino), and differ only in case.